File: Archived2/Java/Silver/24480.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;

public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static int[] visited;
    private static int order = 1;
    private static int startVertex;
    private static ArrayList<ArrayList<Integer>> adj;

    private static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {

        settingInput();

        dfs(startVertex);
        for(int i=1; i<visited.length; i++) {
            sb.append(visited[i]).append("\n");
        }
        System.out.println(sb);
        br.close();
    }

    private static void settingInput() throws IOException {
        int[] initData = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        startVertex = initData[2];
        visited = new int[initData[0] +1];
        adj = new ArrayList<>();
        for (int i=0;i<initData[0]+1;i++) adj.add(new ArrayList<>());
        ArrayList<int[]> edges = new ArrayList<>();
        for (int i=0;i<initData[1];i++) {
            int[] edge = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            edges.add(edge);
            edges.add(new int[] {edge[1], edge[0]});
        }
        edges.sort((a,b) -> {
            if (a[0] == b[0]) return b[1] - a[1];
            else return b[0] - a[0];
        });
        edges.forEach((edge) ->  adj.get(edge[0]).add(edge[1]));
    }

    private static void dfs(int vertex) {
        visited[vertex] = order++;
        adj.get(vertex).forEach((next) -> {
            if (visited[next] == 0) dfs(next);
        });
    }


}

File: Kotlin/Silver/24480.kt
import java.io.BufferedReader
import java.io.InputStreamReader

val br = BufferedReader(InputStreamReader(System.`in`))

lateinit var adj: ArrayList<ArrayList<Int>>
lateinit var visited: IntArray
var startVertex = 0;
var order = 1;

fun main() {

    settingInput()
    dfs(startVertex)
    println( buildString {
        for (i in 1 until visited.size){
            append(visited[i]).append("\n")
        }
    })

}

fun settingInput() {
    val initData = br.readLine().split(" ").map { it.toInt() }
    adj = ArrayList()
    startVertex = initData[2]
    visited = IntArray(initData[0]+1)
    repeat(initData[0]+1) { adj.add(ArrayList()) }
    val edges = ArrayList<IntArray>()
    repeat(initData[1]) {
        val edge = br.readLine().split(" ").map { it.toInt() }.toIntArray()
        edges.add(edge)
        edges.add(intArrayOf(edge[1], edge[0]))
    }
    edges.sortWith( compareByDescending<IntArray> { it[0] }.thenByDescending { it[1] } )
    edges.forEach { adj.get(it[0]).add(it[1]) }
}

fun dfs(vertex: Int) {
    visited[vertex] = order++
    adj.get(vertex).forEach {
        if (visited[it] == 0) dfs(it)
    }
}